### DISCUSSION TITLE
Recieve IDOCs via this Library

### DISCUSSION THREAD

**Discussion by jogibear9988** [dbosoft/YaNco#210] [Q&A] [answered]
We used the sapnco.dll before in our SAP connector.
Within this we had a class like this:

    public class ServerHandler
    {
        [RfcServerFunction(Name = "STFC_CONNECTION")]
        public static void StfcConnection(RfcServerContext context, IRfcFunction function)
        {
            }

        [RfcServerFunction(Name = "IDOC_INBOUND_ASYNCHRONOUS", Default = false)]
        public void IdocInboundAsynchronous(RfcServerContext context, IRfcFunction function)
        {
           
        }
    }

and this functions where then called from SAP Side.
Is this also possible with this library?

And also we send IDOCs with code like this:

        public void SendTestIdoc()
        {
                RfcRepository repo = _sapClient.Repository;
                IRfcFunction bapi = repo.CreateFunction("IDOC_INBOUND_ASYNCHRONOUS");
                var controlTable = bapi.GetTable("IDOC_CONTROL_REC_40");

                IRfcStructure controlStructure = controlTable.Metadata.LineType.CreateStructure();

                controlStructure.SetValue("TABNAM", "EDI_DC40");
                controlStructure.SetValue("MANDT", IdocConfig.MANDT);

                var dataTable = bapi.GetTable("IDOC_DATA_REC_40");

                IRfcStructure dataStructure = dataTable.Metadata.LineType.CreateStructure();

                dataStructure.SetValue("SEGNAM", "E2LINFX");
                dataStructure.SetValue("SDATA", "001Test");

                dataTable.Insert(dataStructure);

                bapi.SetValue("IDOC_DATA_REC_40", dataTable);

                bapi.Invoke(_sapClient);
        }

also possible???

and one more, with the old SAP Library we need to set up multiple config setting:
Don't we need all of them?

ClientConfig:

                { RfcConfigParameters.SystemID,  _sap.ClientConfig.SystemID },
                { RfcConfigParameters.Client,  _sap.ClientConfig.Client },
                { RfcConfigParameters.Language,  _sap.ClientConfig.Language },
                { RfcConfigParameters.PoolSize, "5" },
                { RfcConfigParameters.PeakConnectionsLimit, "10" },
                { RfcConfigParameters.IdleTimeout, "600" },
                { RfcConfigParameters.User, _sap.ClientConfig.User },
                { RfcConfigParameters.Password, _sap.ClientConfig.Password) },
                { RfcConfigParameters.MessageServerHost, _sap.ClientConfig.MessageServerHost);
                { RfcConfigParameters.MessageServerService, _sap.ClientConfig.MessageServerService);
               { RfcConfigParameters.LogonGroup, _sap.ClientConfig.LogonGroup);
                { RfcConfigParameters.AppServerHost, _sap.ClientConfig.AppServerHost);
              {  RfcConfigParameters.SystemNumber,  _sap.ClientConfig.SystemNumber );
              { RfcConfigParameters.GatewayHost, _sap.ClientConfig.GatewayHost);

ServerConfig:

                { RfcConfigParameters.RepositoryDestination,  _sap.ServerConfig.RepositoryDestination },
                { RfcConfigParameters.GatewayHost, _sap.ServerConfig.GatewayHost },
                { RfcConfigParameters.GatewayService, _sap.ServerConfig.GatewayService },
                { RfcConfigParameters.ProgramID,_sap.ServerConfig.ProgramID },
                { RfcConfigParameters.RegistrationCount, "1" }

**Comment by fw2568** [comment]
Hi,

the latest release candidate (https://www.nuget.org/packages/Dbosoft.YaNco/4.3.0-rc.1) adds support for RFC servers. 

A sample can be found here (https://github.com/dbosoft/YaNco/blob/main/test/RfcServerTest/Program.cs). 

Have a look at the samples and if you implement a working idoc receiver I would be happy about a PR with a reference implementation of a idoc receiver. 

The config required is similar as in sapnco as you have to configure always the same settings for a SAP connection. 
A RFC server registers in a SAP gateway and therefore has to specify the connection to it. 
A client is not needed for server only registration, however in that case metadata cannot be retrieved from the sap system and no response can be send back asynchronously. So normally a RFC server will also have a client connection.

Best Regards,
Frank

**Comment by fw2568** [comment, marked as answer]
There is now a sample implementation of a IDoc receiver in this project: https://github.com/dbosoft/YaNco/tree/main/samples/net6.0/ExportMATMAS
